certbot/compat/filesystem.py

```diff
@@ -312,6 +314,116 @@ def is_executable(path):
     return _win_is_executable(path)
 
 
+def has_world_permissions(path):
+    # type: (str) -> bool
+    """
+    Check if everybody/world has any right (read/write/execute) on a file given its path
+    :param str path: path to test
+    :return: True if everybody/world has any right to the file
+    :rtype: bool
+    """
+    if POSIX_MODE:
+        return bool(stat.S_IMODE(os.stat(path).st_mode) & stat.S_IRWXO)
+
+    security = win32security.GetFileSecurity(path, win32security.DACL_SECURITY_INFORMATION)
+    dacl = security.GetSecurityDescriptorDacl()
+
+    return bool(dacl.GetEffectiveRightsFromAcl({
+        'TrusteeForm': win32security.TRUSTEE_IS_SID,
+        'TrusteeType': win32security.TRUSTEE_IS_USER,
+        'Identifier': win32security.ConvertStringSidToSid('S-1-1-0'),
+    }))
+
+
+def compute_private_key_mode(old_key, base_mode):
+    # type: (str, int) -> int
+    """
+    Calculate the POSIX mode to apply to a private key given the previous private key
+    :param str old_key: path to the previous private key
+    :param int base_mode: the minimum modes to apply to a private key
+    :return: the POSIX mode to apply
+    :rtype: int
+    """
+    if POSIX_MODE:
+        # On Linux, we keep read/write/execute permissions
+        # for group and read permissions for everybody.
+        old_mode = (stat.S_IMODE(os.stat(old_key).st_mode) &
+                    (stat.S_IRGRP | stat.S_IWGRP | stat.S_IXGRP | stat.S_IROTH))
+        return base_mode | old_mode
+
+    # On Windows, the mode returned by os.stat is not reliable,
+    # so we do not keep any permission from the previous private key.
+    return base_mode
+
+
+def has_same_ownership(path1, path2):
+    # type: (str, str) -> bool
+    """
+    Return True if the ownership of two files given their respective path is the same
+    :param str path1: path to the first file
```

Comment:
nit: Objects on Windows have a primary group right? If so, I think we should add a line above this saying that this isn't compared.

Reply:
No they do not. Only a owner. I added a comment about that.

Reply:
That's probably right, but do you understand the role of the [primary group](https://docs.microsoft.com/en-us/windows/win32/secauthz/primary-group-of-a-new-object) in the security descriptor? If so, can you explain it to me?

Reply:
Not really. The only other relevant description of that was here: https://casupport.broadcom.com/cadocs/0/CA%203Tera%20Applogic%203%201-ENU/Bookshelf_Files/HTML/RBAC_Guide/index.htm?toc.htm?ManagingNewObjectACLCreation.html

It does not seem related to the primary group definition you have in POSIX, on which you can set permissions for a file. It seems more related to how the default ACL will be created for a new file.

Reply:
Ah, yes, I remember now. It is related to Active Directory. You can find a description here: https://social.technet.microsoft.com/Forums/en-US/c52d6e70-decd-46d4-9431-c4667256edad/what-is-a-primary-group-what-are-its-functoins?forum=winserverDS

In a AD a user is a member of one or more groups. One is the primary one. Out of an Active Directory, local users are member of the Local Group, which the group specific to a local machine.

So here in a Security Descriptor, the primary group is usually the one of the creator. But still it is not related to any effective file permission.

Reply:
I see.

I also found a couple references talking about this [field's usage in the POSIX subsystem](http://www.ntfs.com/ntfs-permissions-security-descriptor.htm), but neither of these things seem particularly relevant to Certbot.